ARCs/arc-0072.md

```diff
@@ -0,0 +1,342 @@
+    },
+    {
+      "name": "transferFrom",
+      "desc": "Transfers ownership of an NFT",
+      "readonly": false,
+      "args": [
+        { "type": "address", "name": "from" },
+        { "type": "address", "name": "to" },
+        { "type": "uint256", "name": "tokenId" }
+      ],
+      "returns": { "type": "void" }
+    },
+  ],
+  "events": [
+    {
+      "name": "Transfer",
+      "desc": "Transfer ownership of an NFT",
+      "args": [
+        {
+          "type": "address",
+          "name": "from",
+          "desc": "The current owner of the NFT"
+        },
+        {
+          "type": "address",
+          "name": "to",
+          "desc": "The new owner of the NFT"
+        },
+        {
+          "type": "uint256",
+          "name": "tokenId",
+          "desc": "The ID of the transferred NFT"
+        }
+      ]
+    }
+  ]
+}
+```
+
+Ownership of a token ID by the zero address indicates that ID is invalid.
+The `ownerOf` method MUST return the zero address for invalid token IDs.
+The `transferFrom` method MUST error when `from` is not the owner of `tokenId`.
+The `transferFrom` method MUST error unless called by the owner of `tokenId` or an approved operator as defined by an extension such as the transfer management extension defined in this ARC.
+The `transferFrom` method MUST emit a `Transfer` event a transfer is successful.
+A `Transfer` event SHOULD be emitted, with `from` being the zero address, when a token is first minted.
+A `Transfer` event SHOULD be emitted, with `to` being the zero address, when a token is destroyed.
+
+All methods in this and other interfaces defined throughout this standard that are marked as `readonly` MUST be read-only as defined by [ARC-22](./arc-0022.md).
+
+The ARC-73 interface selector for this core interface is `0x15974096`.
+
+
+### Metadata Extension
+
+A smart contract NFT that is compliant with this metadata extension MUST implement the interfaces required to comply with the Core NFT Specification, as well as the following interface:
+
+```json
+{
+  "name": "ARC-72 Metadata Extension",
+  "desc": "Smart Contract NFT Metadata Interface",
+  "methods": [
+    {
+      "name": "tokenURI",
+      "desc": "Returns a URI pointing to the NFT metadata",
+      "readonly": true,
+      "args": [
+        { "type": "uint256", "name": "tokenId", "desc": "The ID of the NFT" },
+      ],
+      "returns": { "type": "byte[256]", "desc": "URI to token metadata." }
+    }
+  ],
+}
+```
+
+URIs shorter than the return length MUST be padded with zero bytes at the end of the URI.
+The token URI returned SHOULD be an `ipfs://...` URI so the metadata can't expire or be changed by a lapse or takeover of a DNS registration.
+The token URI SHOULD NOT be an `http://` URI due to security concerns.
+The URI SHOULD resolve to a JSON file following the JSON Metadata File Schema defined in [ARC-3](./arc-0003.md).
+Future standards could define new recommended URI or file formats for metadata.
```

Comment:
This doesn't say what should happen if an invalid tokenId is given. Should we add that into the standard? I imagine people will either cause it to error or return all 0s. It might be nice to enforce one way.

Reply:
If you provide an Invalid TokenId, it will not work, and you will have an error at the app call level. (Key does not exist inside the Box)

Reply:
Sure but you can check if the box exists before accessing it. In fact, we have to do that for the `ownerOf` method call. The spec requires the zero address be returned rather than erroring out. It seems reasonable that someone could follow that pattern here and return a zero byte-string for the URI rather than erroring out. That's why I wondered if we should add into the spec what should happen if the `tokenId` is invalid and remove the ambiguity.